### PR TITLE
Transactions implementation

### PR DESCRIPTION
**Pull request by lucblender**

### CODE REVIEW

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 1]
```diff
@@ -373,6 +371,14 @@ public void removeNode(String nodeName) {
         return (T)internal.nodes.getItem(nodeName).getExternalNode();
     }
 
+    public void beginTransaction(){
+        internal.beginTransaction();
+    }
+
+    public void endTransaction(){
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java]: How about using commitTransaction(), this is what is often used in SQL transactions, so developers might feel more confident about what the method does.

We should add the rollbackTransaction() method which does nothing other than clearing the attribute list...

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 2]
```diff
@@ -540,12 +548,10 @@ public InternalEndpoint(String uuid, CloudioEndpointConfiguration configuration,
...
             String persistenceProvider = configuration.getProperty(MQTT_PERSISTENCE_PROPERTY, MQTT_PERSISTENCE_DEFAULT);
-            if (persistenceProvider.equals(MQTT_PERSISTENCE_MEMORY)) {
-                persistence = new MemoryPersistence();
-            } else if (persistenceProvider.equals(MQTT_PERSISTENCE_FILE)) {
-                persistence = new MqttDefaultFilePersistence();
-            } else if (persistenceProvider.equals(MQTT_PERSISTENCE_NONE)) {
-                persistence = null;
+            if (persistenceProvider.equals(MQTT_PERSISTENCE_TRUE)) {
+                persistence = true;
+            } else if (persistenceProvider.equals(MQTT_PERSISTENCE_FALSE)) {
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java]:
```java
persistence = persistenceProvider.equals(MQTT_PERSISTENCE_TRUE)
```

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 3]
```diff
@@ -935,6 +959,45 @@ private void set(String topic, Stack<String> location, CloudioMessageFormat mess
             }
         }
 
+        private void beginTransaction(){
+            inTransaction = true;
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java]: Protect attribute against concurrent modification

**Review by cm0x4D — CHANGES_REQUESTED** [review]

File: src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java [hunk 4]
```diff
@@ -431,6 +435,7 @@ public void endTransaction(){
         private final CloudioMessageFormat messageFormat;
         private final List<CloudioEndpointListener> listeners = new LinkedList<CloudioEndpointListener>();
         private String jobsFilePath;
+        private final Object transactionLock = new Object();
```
Comment (cm0x4D) [inline src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java]: No need for a dedicated object for `synchronized()`, just do:

```java
synchronized(this){
    inTransaction = //...
}
```

Comment (lucblender) [reply src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java]: wasn't sure if _this_ was already used in some synchronized block, I'll change that then

**Comment by lucblender** [comment]
@cm0x4D I did change the transactionLock in 323b3b9, also did correct the default log level in both endpoint and core in ff3f8df

**Review by cm0x4D — APPROVED** [review]

**Pull request merged**